File: python/mpu9250_driver/ak8963.py
#!/usr/bin/python3
import smbus
import time
import yaml
import numpy as np
from utils.transformation import NED2ENU

WIA = 0x00
INFO = 0x01
ST1 = 0x02
HXL = 0x03
HXH = 0x04
HYL = 0x05
HYH = 0x06
HZL = 0x07
HZH = 0x08
ST2 = 0x09
CNTL = 0x0A
ASAX = 0x10
ASAY = 0x11
ASAZ = 0x12

class AK8963():
    """
    AK8963 I2C driver for acquire magnetometer data

    :param SMBus bus: device I2C port
    :param int address: AK8963 I2C address
    :param str nav_frame: navigation frame
    :param int hz: IMU frequency
    """
    def __init__(self, bus, address, nav_frame="NED", hz=100, calibration=False):
        # I2C connection parameter
        self.bus = bus
        self.address = address

        # Magnetometer parameter
        self.mag_bias = np.zeros((3,1))
        self.mag_scale = np.zeros((3,1))
        self.mag_misalignment = np.zeros((6,1))
        self.mag_strength = 0
        self.calibration = calibration

        # driver parameter
        self.nav_frame = nav_frame # original navigation frame of AK8963 is NED
        self.hz = hz

        # Load old config from yaml file
        if self.calibration == False: 
            f = open("config.yaml", "r")
            self.config = yaml.load(f, Loader=yaml.FullLoader)
            mag_bias = ["mx_bias","my_bias","mz_bias"]
            mag_scale = ["mx_scale","my_scale","mz_scale"]
            # mag_misalignment = ["mag_xy_mis","mag_xz_mis","mag_yx_mis","mag_yz_mis","mag_zx_mis","mag_zy_mis"]
            for i, element in enumerate(mag_bias):
                self.mag_bias[i][0] = self.config[nav_frame][element]
            for i, element in enumerate(mag_scale):
                self.mag_scale[i][0] = self.config[nav_frame][element]
            # for i, element in enumerate(mag_misalignment):
            #     self.mag_misalignment[i][0] = self.config[nav_frame][element]

        # Check parameter
        if (self.nav_frame != "ENU") and (self.nav_frame != "NED"):
            raise ValueError("Navigation frame should be either ENU or NED")
            
    def who_am_i(self):
        """
        Check AK8963 WHOAMI register value
        """
        value = hex(self.bus.read_byte_data(self.address, WIA))
        print("The register value is {}".format(value))
        if value == "0x48":
            print("It is AK8963 default value")
        else:
            print("It is not AK8963 default value")
            raise RuntimeError("AK8963 not found")

    def config_AK8963(self, bit):
        """
        Config AK8963 magnetometer mode

        :param int bit: magnetometer configuration register value
        """
        self.set_mode("fuse rom access", 16)
        self.get_adjust_mag()
        self.set_mode("power down", 16)
        self.set_mode("continuous measure 2", 16)
        self.get_status()

    def get_status(self):
        """
        Check AK8963 magnetometer status
        """
        ST1_value = self.bus.read_byte_data(self.address, ST1)
        bit_0 = ST1_value & int("00000001", 2)
        bit_1 = ST1_value & int("00000010", 2)
        if bit_0 == 0:
            print("Ready in measurement data register or ST2 register")
        elif bit_0 == 1:
            print("Ready in single measurement mode or self-test mode")
        else:
            raise ValueError("AK8963 status 1 register bit 0 error")
        
        if bit_1 == 0:
            print("Ready in measurement data register or ST2 register")
        elif bit_1 == 2:
            print("Data overrun")
        else:
            raise ValueError("AK8963 status 1 register bit 1 error")

    def get_adjust_mag(self):
        """
        AK8963 sensitivity adjustment value for xyz axis
        """
        print("Read sensitivity adjustment value")
        asax = self.bus.read_byte_data(self.address, ASAX)
        asay = self.bus.read_byte_data(self.address, ASAY)
        asaz = self.bus.read_byte_data(self.address, ASAZ)

        self.adjustment_x = (((asax-128)*0.5/128)+1)
        self.adjustment_y = (((asay-128)*0.5/128)+1)
        self.adjustment_z = (((asaz-128)*0.5/128)+1)

    def mag_calibration(self, s: int):
        """
        Calculate the magnetometer bias, scale, misalignment, geomagnetic field strength with four element calibration
        Using least square method to solve the error

        :param int s: time for calibration
        :returns: 
            - mag_scale (ndarray) - 3-axis magnetometer scale (soft-iron offset)
            - mag_bias (ndarray) - 3-axis magnetometer bias (hard-iron offset)
            - mag_misalignment (ndarray) - 3-axis magnetometer misalignment (soft-iron offset)
            - mag_strength (float) - geomagnetic field strength

        .. Reference
        .. [1] `four element calibration <https://www.nxp.com/docs/en/application-note/AN5019.pdf>`
        .. [2] `ten element calibration <https://github.com/nliaudat/magnetometer_calibration/blob/main/calibrate.py>`
        """
        if s > 0 and self.calibration == True:
            input("Please move the IMU in slow motion in all possible directions, the calibration process takes {}s".format(s))
            calibration = []
            target = []
            for i in range(s*self.hz):
                mx, my, mz = self.get_mag()
                calibration.append([mx, my, mz, 1])
                target.append([mx**2+my**2+mz**2])
                time.sleep(1/self.hz)
            calibration = np.array(calibration)
            target = np.array(target)

            error_matrix = np.linalg.inv(calibration.T @ calibration) @ calibration.T @ target

            # calculate bias
            x_bias = 0.5*error_matrix[0][0]
            y_bias = 0.5*error_matrix[1][0]
            z_bias = 0.5*error_matrix[2][0]
            self.mag_bias = np.array([[x_bias],[y_bias],[z_bias]])

            # calculate scale
            self.mag_scale = np.array([[1],[1],[1]])

            # calculate misalignment
            self.mag_misalignment = np.array([[0],[0],[0],[0],[0],[0]])

            # calculate strength
            strength = (error_matrix[3][0]+x_bias**2+y_bias**2+z_bias**2)**0.5
            self.mag_strength = np.array([[strength]])
        return self.mag_scale, self.mag_bias, self.mag_misalignment, self.mag_strength

    def get_mag(self):
        """
        AK8963 magnetometer data in Earth's reference (µT)

        ENU: \n
        mx is positive when the right hand side pointing to north \n
        my is positive when the front side pointing to north \n
        mz is positive when the upper side pointing to north \n

        NED: \n
        mx is positive when the front side pointing to north \n
        my is positive when the right hand side pointing to north \n
        mz is positive when the under side pointing to north \n
    
        :returns: 
            - mx (float) - x-axis magnetometer data in µT
            - my (float) - y-axis magnetometer data in µT
            - mz (float) - z-axis magnetometer data in µT
        """
        try:
            mx = self.read_raw_data(HXH, HXL)*self.mag_fs
            my = self.read_raw_data(HYH, HYL)*self.mag_fs
            mz = self.read_raw_data(HZH, HZL)*self.mag_fs
            # print("raw mx my mz: ", mx, my, mz)
        except:
            raise ConnectionError("I2C Connection Failure")

        # sensitivity adjustment
        mx = mx*self.adjustment_x       
        my = my*self.adjustment_y
        mz = mz*self.adjustment_z

        ST2_value = self.bus.read_byte_data(self.address, ST2)
        bit_3 = ST2_value & int("00001000", 2)
        bit_4 = ST2_value & int("00010000", 2)

        # if bit_3 == 0:
        #     print("Ready in measurement data register or ST2 register")
        # elif bit_3 == 8:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
        #     print("Magnetic sensor overflow occurred")
        # else:
        #     raise ValueError("AK8963 status 2 register bit 3 error")

        # if bit_4 == 0:
        #     print("Magnetic sensor in 14-bit coding")
        # elif bit_4 == 16:
        #     print("Magnetic sensor in 16-bit coding")
        # else:
        #     raise ValueError("AK8963 status 2 register bit 4 error")

        if self.nav_frame == "ENU":
            mx, my, mz = NED2ENU(mx, my, mz)

        # magnetometer model: calibrated measurement = (matrix)*(raw measurement - bias)
        x_scale = self.mag_scale[0][0]
        y_scale = self.mag_scale[1][0]
        z_scale = self.mag_scale[2][0]
        x_bias = self.mag_bias[0][0]
        y_bias = self.mag_bias[1][0]
        z_bias = self.mag_bias[2][0]
        xy_mis = self.mag_misalignment[0][0]
        xz_mis = self.mag_misalignment[1][0]
        yx_mis = self.mag_misalignment[2][0]
        yz_mis = self.mag_misalignment[3][0]
        zx_mis = self.mag_misalignment[4][0]
        zy_mis = self.mag_misalignment[5][0]
        if self.calibration == False:
            mx = (x_scale * mx + xy_mis * my + xz_mis * mz) - x_bias
            my = (yx_mis * mx + y_scale * my + yz_mis * mz) - y_bias
            mz = (zx_mis * mx + zy_mis * my + z_scale * mz) - z_bias
        return mx, my, mz

    def set_mode(self, mode, bit):
        """
        AK8963 CNTL1 register configuration

        :param bool mode: operation mode setting
        :param int bit: output bit setting 
        """
        # power down mode
        if mode == "power down": # 00000000
            value = 0x00
        elif mode == "single measure": # 00000001
            value = 0x01
        elif mode == "continuous measure 1": # 00000010
            value = 0x02 # 8Hz
        elif mode == "continuous measure 2": # 00000110
            value = 0x06 # 100Hz
        elif mode == "external trigger measurement": # 00000100
            value = 0x04
        elif mode == "self test": # 00001000
            value = 0x08
        elif mode == "fuse rom access": # 00001111 
            value = 0x0F
        else:
            raise ValueError("Prohibit mode coding")
        
        if bit == 14:
            value += 0
        elif bit == 16:
            value += 16
        else:
            raise ValueError("Wrong bit coding")

        self.mag_fs = 4912*2/(2**bit)
        print("Set AK8963 to {} mode".format(mode))
        self.bus.write_byte_data(self.address, CNTL, value)
        
    def read_raw_data(self, high_register, low_register):
        """
        Access the high and low registers of the magnetometer to calculate their values into int16 format

        :param int high_register: high registers of the magnetometer
        :param int low_register: low registers of the magnetometer

        :returns: 
            - signed_value (int) - sensor value in int16 format
        """
        high = self.bus.read_byte_data(self.address, high_register)
        low = self.bus.read_byte_data(self.address, low_register)

        # Megre higher bytes and lower bytes data
        unsigned_value = (high << 8) + low

        # Calculate the unsigned int16 range to signed int16 range
        if (unsigned_value >= 32768) and (unsigned_value < 65536):
            signed_value = unsigned_value - 65536
        elif (unsigned_value >= 0) and (unsigned_value < 32768):
            signed_value = unsigned_value
        return signed_value

File: cpp/lib/ak8963.cpp
#include <math.h>
#include <sstream>
#include <vector>
#include "ak8963.h"
#include "transformation.h"
#include "utils.h"

void AK8963::who_am_i()
{
    int value;
    value = this->read_8bit_register(WIA);
    cout << "The register value is 0x" << hex << value << endl;
    if (value == 0x48){
        cout << "It is AK8963 default value" << endl;
    }
    else{
        cout << "It is not AK8963 default value" << endl;
        throw invalid_argument("AK8963 not found");
    }
}

void AK8963::config_AK8963(int mag_parameter=16)
{
    this->set_mode("fuse rom access", mag_parameter);
    this->get_adjust_mag();
    this->set_mode("power down", mag_parameter);
    this->set_mode("continuous measure 2", mag_parameter);
    this->get_status();
}

void AK8963::get_status()
{
    int ST1_value, bit_0, bit_1;
    ST1_value = this->read_8bit_register(ST1);
    bit_0 = ST1_value & 1; // int("00000001", 2)
    bit_1 = ST1_value & 2; // int("00000010", 2)
    if (bit_0 == 0){
        cout << "Ready in measurement data register or ST2 register" << endl;
    }
    else if (bit_0 == 1){
        cout << "Ready in single measurement mode or self-test mode" << endl;
    }
    else{
        throw invalid_argument("AK8963 status 1 register bit 0 error");
    }

    if (bit_1 == 0){
        cout << "Ready in measurement data register or ST2 register" << endl;
    }
    else if (bit_1 == 2){
        cout << "Data overrun" << endl;
    }
    else{
        throw invalid_argument("AK8963 status 1 register bit 1 error");
    }
}

void AK8963::get_adjust_mag()
{
    int asax, asay, asaz;
    cout << "Read sensitivity adjustment value" << endl;
    asax = this->read_8bit_register(ASAX);
    asay = this->read_8bit_register(ASAY);
    asaz = this->read_8bit_register(ASAZ);

    this->adjustment_x = (((asax-128)*0.5/128)+1);
    this->adjustment_y = (((asay-128)*0.5/128)+1);
    this->adjustment_z = (((asaz-128)*0.5/128)+1);
}

mag_calib AK8963::mag_calibration(float s)
{
    float mx, my, mz;
    Vector3d mag;
    Matrix4d error_matrix;
    Vector4d error_vector;
    if ((s > 0) and (this->calibration == true))
    {
        stringstream ss;
        Vector3d total_gyro, current_gyro;
        total_gyro << 0, 0, 0;
        ss << "Please move the IMU in slow motion in all possible directions, the calibration process takes " << s << "s";
        cout << ss.str();
        cin.ignore(); // Press enter with nothing
        MatrixXd calibration = MatrixXd(int(s*this->hz),4);
        VectorXd target = VectorXd(int(s*this->hz));
        for (int i=0;i<(s*this->hz);i++)
        {
            mag = this->get_mag();
            mx = mag[0];
            my = mag[1];
            mz = mag[2];

            calibration(i,0) = mx;
            calibration(i,1) = my;
            calibration(i,2) = mz;
            calibration(i,3) = 1;
            target[i] = pow(mx,2)+pow(my,2)+pow(mz,2);
            time_sleep(0, 1/this->hz);
        }
        
        error_matrix = calibration.transpose() * calibration;
        error_vector = error_matrix.inverse().eval() * calibration.transpose() * target;

        this->mx_bias = 0.5*error_vector[0];
        this->my_bias = 0.5*error_vector[1];
        this->mz_bias = 0.5*error_vector[2];
        
        this->mag_bias << this->mx_bias, this->my_bias, this->mz_bias;
        this->mag_scale << this->mx_scale, this->my_scale, this->mz_scale;
        this->mag_misalignment << this->mxy_mis, this->mxz_mis, this->myx_mis, this->myz_mis, this->mzx_mis, this->mzy_mis;
        this->mag_strength = pow((error_vector[3]+pow(this->mx_bias,2)+pow(this->my_bias,2)+pow(this->mz_bias,2)),0.5);
    }
    mag_calib result;
    result.scale = this->mag_scale;
    result.bias = this->mag_bias;
    result.misalignment = this->mag_misalignment;
    result.strength = this->mag_strength;
    cout << "Finish magnetometer calibration" << endl;
    return result;
}

Vector3d AK8963::get_mag()
{
    Vector3d mag;
    int ST2_value, bit_3, bit_4;
    float mx, my, mz;

    mx = this->read_raw_data(HXH, HXL)*this->mag_fs;
    my = this->read_raw_data(HYH, HYL)*this->mag_fs;
    mz = this->read_raw_data(HZH, HZL)*this->mag_fs;

    // sensitivity adjustment
    mx = mx*this->adjustment_x;
    my = my*this->adjustment_y;
    mz = mz*this->adjustment_z;

    ST2_value = this->read_8bit_register(ST2);
    bit_3 = ST2_value & 8; // int("00001000", 2)
    bit_4 = ST2_value & 16; // int("00010000", 2)

    // if (bit_3 == 0){
    //     cout << "Ready in measurement data register or ST2 register" << endl;
    // }
    // else if (bit_3 == 8){
    //     cout << "Magnetic sensor overflow occurred" << endl;
    // }
    // else{
    //     throw invalid_argument("AK8963 status 2 register bit 3 error");    
    // }

    // if (bit_4 == 0){
    //     cout << "Magnetic sensor in 14-bit coding" << endl;
    // }
    // else if (bit_4 == 16){
    //     cout << "Magnetic sensor in 16-bit coding" << endl;
    // }
    // else{
    //     throw invalid_argument("AK8963 status 2 register bit 4 error");    
    // }

    if (this->nav_frame == "ENU")
    {
        mag = NED2ENU(mx, my, mz);
        mx = mag[0];
        my = mag[1];
        mz = mag[2]; 
    }

    // magnetometer model: calibrated measurement = (matrix)*(raw measurement - bias)
    if (this->calibration == false)
    {
        mx = (this->mx_scale * mx + this->mxy_mis * my + this->mxz_mis * mz) - this->mx_bias;
        my = (this->myx_mis * mx + this->my_scale * my + this->myz_mis * mz) - this->my_bias;
        mz = (this->mzx_mis * mx + this->mzy_mis * my + this->mz_scale * mz) - this->mz_bias;
    }

    mag << mx, my, mz;
    return mag;
}

void AK8963::set_mode(string mode, int bit)
{
    int value;
    if (mode == "power down"){
        value = 0x00; // 00000000
    }
    else if (mode == "single measure"){
        value = 0x01; // 00000001
    }
    else if (mode == "continuous measure 1"){
        value = 0x02; // 8Hz, 00000010
    }
    else if (mode == "continuous measure 2"){
        value = 0x06; // 100Hz, 00000110
    }
    else if (mode == "external trigger measurement"){
        value = 0x04; // 8Hz, 00000100
    }
    else if (mode == "self test"){
        value = 0x08; // 00001000
    }
    else if (mode == "fuse rom access"){
        value = 0x0F; // 00001111
    }
    else{
        throw invalid_argument("Prohibit mode coding");        
    }

    if (bit == 14){
        value += 0;
    }
    else if (bit == 16){
        value += 16;
    }
    else{
        throw invalid_argument("Wrong bit coding");
    }

    this->mag_fs = 4912*2/pow(2, bit);
    cout << "Set AK8963 to " << mode << " mode" << endl;
    wiringPiI2CWriteReg8(this->i2c, CNTL, value);
}

int AK8963::read_raw_data(int high_register, int low_register)
{
    int high=-1, low=-1, unsigned_value, signed_value;
    while (high == -1) {
        high = wiringPiI2CReadReg8(this->i2c, high_register);
        if (high != -1) {
            break;
        }
    }
    while (low == -1) {
        low = wiringPiI2CReadReg8(this->i2c, low_register);
        if (low != -1) {
            break;
        }
    }

    // Megre higher bytes and lower bytes data
    unsigned_value = (high << 8) + low;

    // Calculate the unsigned int16 range to signed int16 range
    if ((unsigned_value >= 32768) and (unsigned_value < 65536)){
        signed_value = -(~(unsigned_value - 65536)+1);
    }
    else if ((unsigned_value >= 0) and (unsigned_value < 32768)){
        signed_value = unsigned_value;
    }
    return signed_value;
}

int AK8963::read_8bit_register(int single_register)
{
    int data=-1;
    while (data == -1) {
        data = wiringPiI2CReadReg8(this->i2c, single_register);
        if (data != -1) {
            break;
            return data;
        }
    }
}